openprescribing/frontend/management/commands/measure_sql/vw__opioids_total_dmd.sql

```diff
@@ -0,0 +1,89 @@
+-- This SQL is checked in to the git repo at measure_sql/vw__opioids_total_dmd.sql.
+-- Do not make changes directly in BQ!  Instead, change the version in the repo and run
+--
+--     ./manage.py create_bq_measure_views
+
+#subquery to create a "simple" administration route. 
+WITH 
+simp_form AS ( 
+SELECT DISTINCT vmp, #vmp code
+                CASE 
+                    WHEN descr LIKE '%injection%' THEN 'injection' #creates "injection" as route, regardless of whether injection or infusion. this also removes injection routes, e.g.
+                    WHEN descr LIKE '%infusion%' THEN 'injection'  #s/c, i/v etc, AS often injections have many licensed routes, which would multiply the row
+                    ELSE SUBSTR(form.descr, STRPOS(form.descr,".")+1) #takes the dosage form out of the string (e.g. tablet.oral) TO leave route.
+                END AS simple_form 
+FROM dmd.ont AS ont #the coded route for dosage form, includes vmp code 
+INNER JOIN dmd.ontformroute AS form ON form.cd=ont.form #text description of route 
+  )
+
+#subquery to normalise strength to mg
+,norm_vpi AS (
+SELECT vmp, #vmp code
+       ing, #ing code
+       strnt_nmrtr_val, #numerator strength value
+       strnt_nmrtr_uom, #numerator unit of measurement
+       unit_num.descr AS unit_num, #numerator unit 
+       unit_den.descr AS unit_den, #denominator unit
+       CASE
+           WHEN unit_num.descr = 'microgram' THEN vpi.strnt_nmrtr_val / 1000 #creates miligram value from mcg value
+           WHEN unit_num.descr = 'gram' THEN vpi.strnt_nmrtr_val * 1000 #creates miligram value from gram value
+           ELSE vpi.strnt_nmrtr_val
+       END AS strnt_nmrtr_val_mg, #all listed drugs now in miligram rather than g or mcg
+       CASE
+           WHEN unit_den.descr = 'litre' THEN vpi.strnt_dnmtr_val * 1000 #some denominators listed as litre, so create mililitre value
+           ELSE vpi.strnt_dnmtr_val
+       END AS strnt_dnmtr_val_ml #denominator now in ml
+FROM dmd.vpi AS vpi
+LEFT JOIN dmd.unitofmeasure AS unit_num ON vpi.strnt_nmrtr_uom = unit_num.cd #join to create text value for numerator unit
+LEFT JOIN dmd.unitofmeasure AS unit_den ON vpi.strnt_dnmtr_uom = unit_den.cd) #join to create text value for denominator unit
+
+#subquery to create single BNF table for AMPs and VMPs
```

Comment:
Worth being clear that this contains one row for each combination of (VMP ID, BNF code, name)?

Reply:
But is that what you want?

Reply:
You're going to have the same BNF code appearing multiple times in this subquery.

In the main query, you're selecting from `normalised_prescribing rx INNER JOIN vmp_amp vmp ON rx.bnf_code = vmp.bnf_code`, which means you'll have duplicate rows in your output.

```
SELECT * FROM measures.vw__opioid_measure_dmd WHERE month = '2020-09-01' AND practice = 'A81001' AND bnf_code = '0407010F0AAAAAA'
```


Row | month | practice | pct | strnt_dnmtr_val_ml | quantity | id | nm | vmp_nm | bnf_code | strnt_nmrtr_val_mg | ome_dose | ome |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
1 | 2020-09-01 00:00:00 UTC | A81001 | 16C | null | 1214.0 | 261000 | Codeine phosphate | Co-codamol 8mg/500mg tablets | 0407010F0AAAAAA | 8 | 971.2 | 0.1 |  
2 | 2020-09-01 00:00:00 UTC | A81001 | 16C | null | 1214.0 | 261000 | Codeine phosphate | Co-codamol 8mg/500mg caplets | 0407010F0AAAAAA | 8 | 971.2 | 0.1 |  

This will mean that some records will be double counted in the measure.

Reply:
@inglesp just reviewing this after a few months.  This duplication can be resolved by changing the source of the name from having `vmp_nm` to `bnf_nm` from the `normalised_prescribing` table.  It's mainly a problem with caplets and tablets, but there is only ever one BNF name per presentation.

Reply:
| Row | month                   | practice | pct | strnt_dnmtr_val_ml | quantity | id     | nm                | bnf_nm                       | bnf_code        | strnt_nmrtr_val_mg | ome_dose | ome |
|-----|-------------------------|----------|-----|--------------------|----------|--------|-------------------|------------------------------|-----------------|--------------------|----------|-----|
| 1   | 2020-09-01 00:00:00 UTC | A81001   | 16C |               null |   2428.0 | 261000 | Codeine phosphate | Co-codamol 8mg/500mg tablets | 0407010F0AAAAAA |                  8 |   1942.4 | 0.1 |

Reply:
ah, although does that just double the quantity...

Reply:
Returned to this after a long time...

Fixed the duplication by using concatenation of `bnf_code` to link to a VMP for all drugs, rather than having separate VMP and AMP list:

```
#subquery to create a "simple" administration route. 
WITH 
simp_form AS ( 
SELECT DISTINCT vmp, #vmp code
                CASE 
                    WHEN descr LIKE '%injection%' THEN 'injection' #creates "injection" as route, regardless of whether injection or infusion. this also removes injection routes, e.g.
                    WHEN descr LIKE '%infusion%' THEN 'injection'  #s/c, i/v etc, AS often injections have many licensed routes, which would multiply the row
                    ELSE SUBSTR(form.descr, STRPOS(form.descr,".")+1) #takes the dosage form out of the string (e.g. tablet.oral) TO leave route.
                END AS simple_form 
FROM dmd.ont AS ont #the coded route for dosage form, includes vmp code 
INNER JOIN dmd.ontformroute AS form ON form.cd=ont.form #text description of route 
  )

#subquery to normalise strength to mg
,norm_vpi AS (
SELECT vmp, #vmp code
       ing, #ing code
       strnt_nmrtr_val, #numerator strength value
       strnt_nmrtr_uom, #numerator unit of measurement
       unit_num.descr AS unit_num, #numerator unit 
       unit_den.descr AS unit_den, #denominator unit
       CASE
           WHEN unit_num.descr = 'microgram' THEN vpi.strnt_nmrtr_val / 1000 #creates miligram value from mcg value
           WHEN unit_num.descr = 'gram' THEN vpi.strnt_nmrtr_val * 1000 #creates miligram value from gram value
           ELSE vpi.strnt_nmrtr_val
       END AS strnt_nmrtr_val_mg, #all listed drugs now in miligram rather than g or mcg
       CASE
           WHEN unit_den.descr = 'litre' THEN vpi.strnt_dnmtr_val * 1000 #some denominators listed as litre, so create mililitre value
           ELSE vpi.strnt_dnmtr_val
       END AS strnt_dnmtr_val_ml #denominator now in ml
FROM dmd.vpi AS vpi
LEFT JOIN dmd.unitofmeasure AS unit_num ON vpi.strnt_nmrtr_uom = unit_num.cd #join to create text value for numerator unit
LEFT JOIN dmd.unitofmeasure AS unit_den ON vpi.strnt_dnmtr_uom = unit_den.cd) #join to create text value for denominator unit
  
#main query to calculate the OME
SELECT rx.month, 
       rx.practice, 
       rx.pct, 
       vpi.strnt_dnmtr_val_ml,
       sum(rx.quantity) as quantity,
       ing.id, #ingredient DM+D code. Combination products will have more than one ing code per VMP, e.g. co-codamol will have ing for paracetamoland codeine
       ing.nm, #ingredient name vmp.bnf_code AS bnf_code,
       #vmp.nm AS vmp_nm, #VMP name
       vmp.bnf_code as bnf_code, #BNF code to link to prescribing data
       rx.bnf_name as bnf_name, #BNF name from prescribing data
       vpi.strnt_nmrtr_val_mg, #strength numerator in mg
       SUM(quantity*ome*(CASE
           WHEN ing.id=373492002 AND form.simple_form = 'transdermal' THEN (vpi.strnt_nmrtr_val_mg*72)/coalesce(vpi.strnt_dnmtr_val_ml, 1) # creates 72 hour dose for fentanyl transdermal patches, as doses are per hour on DM+D)
           WHEN ing.id=387173000 AND form.simple_form = 'transdermal' AND vpi.strnt_nmrtr_val IN (5, 10, 15, 20) THEN (vpi.strnt_nmrtr_val_mg*168)/coalesce(vpi.strnt_dnmtr_val_ml, 1) # creates 168 hour (7 day) dose for low-dose buprenorphine patch
           WHEN ing.id=387173000 AND form.simple_form = 'transdermal' AND vpi.strnt_nmrtr_val IN (35, 52.5, 70) THEN (vpi.strnt_nmrtr_val_mg*96)/coalesce(vpi.strnt_dnmtr_val_ml, 1) # creates 96 hour dose for higher-dose buprenorphine patch
           ELSE strnt_nmrtr_val_mg/coalesce(vpi.strnt_dnmtr_val_ml, 1) #all other products have usual dose - coalesce as solid dose forms do not have a denominator
       END)) AS ome_dose, 
       opioid.ome AS ome
FROM norm_vpi AS vpi #VPI has both ING and VMP codes in the table
INNER JOIN dmd.ing AS ing ON vpi.ing=ing.id #join to ING to get ING codes and name
INNER JOIN dmd.vmp AS vmp ON vpi.vmp=vmp.id #join to get BNF codes for both VMPs and AMPs joined indirectly TO ING. 
INNER JOIN simp_form AS form ON vmp.id=form.vmp #join to subquery for simplified administration route
INNER JOIN richard.opioid_class AS opioid ON opioid.id=ing.id AND opioid.form=form.simple_form #join to OME table, which has OME value for ING/route pairs
INNER JOIN hscic.normalised_prescribing AS rx ON CONCAT(SUBSTR(rx.bnf_code,0,9),'AA',SUBSTR(rx.bnf_code,-2,2)) = CONCAT(SUBSTR(vmp.bnf_code,0,11),SUBSTR(vmp.bnf_code,-2,2))
WHERE rx.bnf_code NOT LIKE '0410%' #remove drugs used in opiate dependence
GROUP BY rx.month, 
         rx.practice, 
         rx.pct,
         id,
         ing.nm,
         vmp.bnf_code,
         rx.bnf_name,
         vpi.strnt_nmrtr_val,
         strnt_nmrtr_val_mg,
         vpi.strnt_dnmtr_val_ml,
         opioid.ome
```

Reply:
tidied version:
```
#subquery to create a "simple" administration route. 
WITH simp_form AS (
  SELECT 
    vmp, #vmp code
    CASE WHEN descr LIKE '%injection%' THEN 'injection' #creates "injection" as route, regardless of whether injection or infusion. this also removes injection routes, e.g.
    WHEN descr LIKE '%infusion%' THEN 'injection' #s/c, i/v etc, AS often injections have many licensed routes, e.g "solutioninjection.subcutaneous" AND solutioninjection.intramuscular"which would multiply the row
    ELSE SUBSTR(
      form.descr, 
      STRPOS(form.descr, ".")+ 1) #takes the dosage form out of the string (e.g. tablet.oral) TO leave route.
    END AS simple_form 
  FROM 
    dmd.ont AS ont #the coded route for dosage form, includes vmp code 
    INNER JOIN dmd.ontformroute AS form ON form.cd = ont.form #text description of route 
    )


#subquery to normalise strength to mg
,norm_vpi AS (
    SELECT 
    vmp, #vmp code
    ing, #ing code
    strnt_nmrtr_val,#numerator strength value
    strnt_nmrtr_uom,#numerator unit of measurement
    unit_num.descr as num_unit, #numerator unit 
    unit_den.descr as den_unit, #denominator unit
    CASE WHEN unit_num.descr = 'microgram' THEN vpi.strnt_nmrtr_val / 1000 #creates miligram value from mcg value
    WHEN unit_num.descr = 'gram' THEN vpi.strnt_nmrtr_val * 1000 #creates miligram value from gram value
    WHEN unit_num.descr = 'mg' THEN vpi.strnt_nmrtr_val #no change if mg value
    ELSE NULL # will give a null value if a non-stanard dosage unit - this can then be checked if neccesary
    END AS strnt_nmrtr_val_mg, #all listed drugs now in miligram rather than g or mcg
    CASE WHEN unit_den.descr = 'litre' THEN vpi.strnt_dnmtr_val * 1000 #some denominators listed as litre, so create mililitre value
    WHEN unit_den.descr = 'ml' THEN vpi.strnt_dnmtr_val #no change if mililitre value
    ELSE NULL # will give a null value if a non-stanard dosage unit - this can then be checked if neccesary
    END AS strnt_dnmtr_val_ml #denominator now in ml
    FROM 
    dmd.vpi AS vpi 
    LEFT JOIN dmd.unitofmeasure AS unit_num ON vpi.strnt_nmrtr_uom = unit_num.cd #join to create text value for numerator unit
    LEFT JOIN dmd.unitofmeasure AS unit_den ON vpi.strnt_dnmtr_uom = unit_den.cd #join to create text value for denominator unit
) 
 
#main query to calculate the OME
SELECT 
  rx.month, 
  rx.practice, 
  rx.pct, 
  vpi.strnt_dnmtr_val_ml, 
  sum(rx.quantity) as quantity, 
  ing.id, #ingredient DM+D code. Combination products will have more than one ing code per VMP, e.g. co-codamol will have ing for paracetamoland codeine
  ing.nm,#ingredient name
  vmp.bnf_code as bnf_code, #BNF code to link to prescribing data
  rx.bnf_name as bnf_name, #BNF name from prescribing data
  vpi.strnt_nmrtr_val_mg, #strength numerator in mg
  SUM(
    quantity * ome *(
      CASE WHEN ing.id = 373492002 
      AND form.simple_form = 'transdermal' THEN (vpi.strnt_nmrtr_val_mg * 72)/ coalesce(vpi.strnt_dnmtr_val_ml, 1) # creates 72 hour dose for fentanyl transdermal patches, as doses are per hour on DM+D)
      WHEN ing.id = 387173000 
      AND form.simple_form = 'transdermal' 
      AND vpi.strnt_nmrtr_val IN (5, 10, 15, 20) THEN (vpi.strnt_nmrtr_val_mg * 168)/ coalesce(vpi.strnt_dnmtr_val_ml, 1) # creates 168 hour (7 day) dose for low-dose buprenorphine patch
      WHEN ing.id = 387173000 
      AND form.simple_form = 'transdermal' 
      AND vpi.strnt_nmrtr_val IN (35, 52.5, 70) THEN (vpi.strnt_nmrtr_val_mg * 96)/ coalesce(vpi.strnt_dnmtr_val_ml, 1) # creates 96 hour dose for higher-dose buprenorphine patch
      ELSE strnt_nmrtr_val_mg / coalesce(vpi.strnt_dnmtr_val_ml, 1) #all other products have usual dose - coalesce as solid dose forms do not have a denominator
      END
    )
  ) AS ome_dose, 
  opioid.ome AS ome 
FROM 
  norm_vpi AS vpi #VPI has both ING and VMP codes in the table
  INNER JOIN dmd.ing AS ing ON vpi.ing = ing.id #join to ING to get ING codes and name
  INNER JOIN dmd.vmp AS vmp ON vpi.vmp = vmp.id #join to get BNF codes for both VMPs and AMPs joined indirectly TO ING. 
  INNER JOIN simp_form AS form ON vmp.id = form.vmp #join to subquery for simplified administration route
  INNER JOIN richard.opioid_class AS opioid ON opioid.id = ing.id AND opioid.form = form.simple_form #join to OME table, which has OME value for ING/route pairs
  INNER JOIN hscic.normalised_prescribing AS rx ON CONCAT(
    SUBSTR(rx.bnf_code, 0, 9), 
    'AA', 
    SUBSTR(rx.bnf_code,-2, 2)
  ) = CONCAT(
    SUBSTR(vmp.bnf_code, 0, 11), 
    SUBSTR(vmp.bnf_code,-2, 2)
  ) #uses bnf code structure to join both branded and generic prescribing data to generic VMP codes - which stops chance of duplication of VMP/AMP names
WHERE 
  rx.bnf_code NOT LIKE '0410%' #remove drugs used in opiate dependence
GROUP BY 
  rx.month, 
  rx.practice, 
  rx.pct, 
  id, 
  ing.nm, 
  vmp.bnf_code, 
  rx.bnf_name, 
  vpi.strnt_nmrtr_val, 
  strnt_nmrtr_val_mg, 
  vpi.strnt_dnmtr_val_ml, 
  opioid.ome
```